File: src/discover_inputs.py
#!/usr/bin/env python3
"""Discover and sort all input documents chronologically.

Scans docs/converted/ for meetings and emails, extracts dates
from filenames, and returns a sorted list for processing.

Skips non-content files (test emails, calendar invites, happening-now notifications).
"""

import json
import re
import sys
from pathlib import Path

SKIP_PATTERNS = [
    r'teszt-',
    r'calendar-invite-',
    r'happening-now-',
    r'invitation-for-',
]

DATE_PATTERN = re.compile(r'(\d{4}-\d{2}-\d{2})')


def extract_date(filename: str) -> str | None:
    """Extract YYYY-MM-DD date from filename."""
    m = DATE_PATTERN.search(filename)
    return m.group(1) if m else None


def should_skip(filename: str) -> bool:
    """Check if file should be skipped based on filename patterns."""
    lower = filename.lower()
    return any(re.search(p, lower) for p in SKIP_PATTERNS)


def classify_document(path: Path) -> str:
    """Classify document type from its path."""
    parts = path.parts
    if 'meetings' in parts:
        return 'meeting'
    if 'emails' in parts:
        return 'email'
    if 'discord' in parts:
        return 'discord'
    if 'discord-voice' in parts:
        return 'discord-voice'
    if 'client-spec' in parts:
        return 'client-spec'
    return 'other'


def discover(base_dir: Path) -> list[dict]:
    """Find all input documents and sort chronologically."""
    results = []

    converted_dir = base_dir / 'docs' / 'converted'
    if not converted_dir.exists():
        print(f"Error: {converted_dir} not found", file=sys.stderr)
        sys.exit(1)

    for md_file in sorted(converted_dir.rglob('*.md')):
        filename = md_file.name

        if should_skip(filename):
            continue

        doc_type = classify_document(md_file)
        date = extract_date(filename)
        rel_path = str(md_file.relative_to(base_dir))

        results.append({
            'path': rel_path,
            'filename': filename,
            'type': doc_type,
            'date': date or '0000-00-00',
        })

    results.sort(key=lambda d: (d['date'], d['path']))

    return results


def main():
    base_dir = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('.')

    documents = discover(base_dir)

    summary = {}
    for doc in documents:
        t = doc['type']
        summary[t] = summary.get(t, 0) + 1

    output = {
        'base_dir': str(base_dir),
        'total': len(documents),
        'by_type': summary,
        'documents': documents,
    }

    json.dump(output, sys.stdout, ensure_ascii=False, indent=2)


if __name__ == '__main__':
    main()

File: src/parse/discover-inputs.ts
import * as fs from "node:fs";
import * as path from "node:path";

const SKIP_PATTERNS: string[] = [
  "teszt-",
  "calendar-invite-",
  "happening-now-",
  "invitation-for-",
];

const DATE_PATTERN = /(\d{4}-\d{2}-\d{2})/;

export interface Document {
  path: string;
  filename: string;
  type: string;
  date: string;
}

export interface DiscoverResult {
  base_dir: string;
  total: number;
  by_type: Record<string, number>;
  documents: Document[];
}

export function extractDate(filename: string): string | null {
  const m = DATE_PATTERN.exec(filename);
  return m ? m[1] : null;
}

export function shouldSkip(filename: string): boolean {
  const lower = filename.toLowerCase();
  return SKIP_PATTERNS.some((p) => new RegExp(p).test(lower));
}

export function classifyDocument(filePath: string): string {
  const parts = filePath.split(path.sep);
  if (parts.includes("meetings")) return "meeting";
  if (parts.includes("emails")) return "email";
  if (parts.includes("discord")) return "discord";
  if (parts.includes("discord-voice")) return "discord-voice";
  if (parts.includes("client-spec")) return "client-spec";
  return "other";
}

function walkMdFiles(dir: string): string[] {
  const results: string[] = [];
  if (!fs.existsSync(dir)) return results;

  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...walkMdFiles(fullPath));
    } else if (entry.isFile() && entry.name.endsWith(".md")) {
      results.push(fullPath);
    }
  }
  return results.sort();
}

export function discover(baseDir: string): Document[] {
  const convertedDir = path.join(baseDir, "docs", "converted");
  if (!fs.existsSync(convertedDir)) {
    process.stderr.write(`Error: ${convertedDir} not found\n`);
    process.exit(1);
  }

  const mdFiles = walkMdFiles(convertedDir);
  const results: Document[] = [];

  for (const mdFile of mdFiles) {
    const filename = path.basename(mdFile);

    if (shouldSkip(filename)) continue;

    const docType = classifyDocument(mdFile);
    const date = extractDate(filename);
    const relPath = path.relative(baseDir, mdFile);

    results.push({
      path: relPath,
      filename,
      type: docType,
      date: date ?? "0000-00-00",
    });
  }

  results.sort((a, b) => {
    if (a.date !== b.date) return a.date < b.date ? -1 : 1;
    return a.path < b.path ? -1 : a.path > b.path ? 1 : 0;
  });

  return results;
}

function main(): void {
  const baseDir = process.argv[2] ?? ".";

  const documents = discover(baseDir);

  const summary: Record<string, number> = {};
  for (const doc of documents) {
    summary[doc.type] = (summary[doc.type] ?? 0) + 1;
  }

  const output: DiscoverResult = {
    base_dir: baseDir,
    total: documents.length,
    by_type: summary,
    documents,
  };

  process.stdout.write(JSON.stringify(output, null, 2));
}

const isMain =
  process.argv[1] &&
  fs.realpathSync(process.argv[1]) === fs.realpathSync(import.meta.url.replace("file://", ""));
if (isMain) {
  main();
}
